homeassistant/components/sonos/sensor.py

```diff
@@ -0,0 +1,213 @@
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.icon import icon_for_battery_level
+
+from . import SonosData
+from .const import (
+    BATTERY_SCAN_INTERVAL,
+    DATA_SONOS,
+    SONOS_DISCOVERY_UPDATE,
+    SONOS_PROPERTIES_UPDATE,
+    SONOS_SEEN,
+    SONOS_UNSEEN,
+)
+from .entity import SonosEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_BATTERY_LEVEL = "battery_level"
+ATTR_BATTERY_CHARGING = "charging"
+ATTR_BATTERY_POWERSOURCE = "power_source"
+
+EVENT_CHARGING = {
+    "CHARGING": True,
+    "NOT_CHARGING": False,
+}
+
+
+def fetch_battery_info_or_none(soco: SoCo) -> Optional[Dict[str, Any]]:
+    """Fetch battery_info from the given SoCo object.
+
+    Returns None if the device doesn't support battery info
+    or if the device is offline.
+    """
+    with contextlib.suppress(ConnectionError, TimeoutError, SoCoException):
+        return soco.get_battery_info()
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Sonos from a config entry."""
+
+    sonos_data = hass.data[DATA_SONOS]
+
+    async def _async_create_entity(soco: SoCo) -> Optional[SonosBatteryEntity]:
+        if battery_info := await hass.async_add_executor_job(
+            fetch_battery_info_or_none, soco
+        ):
+            return SonosBatteryEntity(soco, sonos_data, battery_info)
+        return None
+
+    async def _async_create_entities(soco: SoCo):
+        if entity := await _async_create_entity(soco):
+            async_add_entities([entity])
+
+    async_dispatcher_connect(hass, SONOS_DISCOVERY_UPDATE, _async_create_entities)
+
+    entities = []
+    # create any entities for devices that exist already
+    for soco in sonos_data.discovered.values():
+        if entity := await _async_create_entity(soco):
+            entities.append(entity)
+
+    if entities:
+        async_add_entities(entities)
+
+
+class SonosBatteryEntity(SonosEntity, Entity):
+    """Representation of a Sonos Battery entity."""
+
+    def __init__(self, soco: SoCo, sonos_data: SonosData, battery_info: Dict[str, Any]):
+        """Initialize a SonosBatteryEntity."""
+        super().__init__(soco, sonos_data)
+        self._battery_info = battery_info
+
+    async def async_added_to_hass(self) -> None:
+        """Register polling callback when added to hass."""
+        self.async_on_remove(
+            self.hass.helpers.event.async_track_time_interval(
+                self.update, BATTERY_SCAN_INTERVAL
+            )
+        )
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, f"{SONOS_SEEN}-{self.soco.uid}", self.async_seen
+            )
+        )
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, f"{SONOS_UNSEEN}-{self.soco.uid}", self.async_unseen
+            )
+        )
```

Comment:
Since every entity is expecting to implement `async_seen` and `async_unseen`, we could make it an abstract method in the base class move these to the base class and do `await super().async_added_to_hass()`